tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -175,115 +211,123 @@ tfw_sched_hash_get_sg_conn(TfwMsg *msg, TfwSrvGroup *sg)
 static TfwSrvConn *
 tfw_sched_hash_get_srv_conn(TfwMsg *msg, TfwServer *srv)
 {
-	unsigned long msg_hash;
-	size_t tries;
-	TfwHashSrv *srv_cl = srv->sched_data;
+	TfwHashConnList *cl = srv->sched_data;
 
 	/*
-	 * For @srv without connections srv_cl will be NULL, that normally
+	 * For @srv without connections @cl will be NULL, that normally
 	 * does not happen in real life, but unit tests check that case.
 	*/
-	if (unlikely(!srv_cl))
+	if (unlikely(!cl))
 		return NULL;
 
-	msg_hash = tfw_http_req_key_calc((TfwHttpReq *)msg);
-	/* Try several times even if server has just a few connections. */
-	tries = srv_cl->conn_n + 1;
-	while (--tries) {
-		unsigned long best_weight = 0;
-		TfwSrvConn *best_conn = NULL;
-
-		__find_best_conn(&best_conn, srv_cl, &best_weight, msg_hash);
-		if (unlikely(!best_conn))
-			return NULL;
-		if (likely(!tfw_srv_conn_restricted(best_conn)
-			   && !tfw_srv_conn_queue_full(best_conn)
-			   && tfw_srv_conn_get_if_live(best_conn)))
-			return best_conn;
-	}
-	return NULL;
+	return __find_best_conn(msg, cl);
 }
 
 static void
 tfw_sched_hash_del_grp(TfwSrvGroup *sg)
 {
-	size_t si;
-	TfwHashSrvList *sl = sg->sched_data;
+	TfwServer *srv;
+
+	list_for_each_entry(srv, &sg->srv_list, list)
+		srv->sched_data = NULL;
+
+	if (sg->sched_data) {
+		kfree(sg->sched_data);
+		sg->sched_data = NULL;
+	}
+}
+
+static int
+__add_conn(TfwHashConnList *cl, TfwSrvConn *conn, unsigned long hash)
+{
+	ssize_t idx;
+	TfwHashConn new_hcon = {hash, conn};
+
+	/* @cl is empty. */
+	if (unlikely(!cl->conn_n)) {
+		cl->conns[0] = new_hcon;
+		++cl->conn_n;
+
+		return 0;
+	}
+
+	idx = __bsearch(hash, cl);
+	if (cl->conns[idx].hash == hash)
+		return -EEXIST;
+
+	/* Need to insert connection before or after found index. */
+	if ((hash > cl->conns[idx].hash) && (idx != (ssize_t)cl->conn_n))
+		++idx;
+
+	if ((size_t)idx != cl->conn_n)
+		memmove(&cl->conns[idx+1], &cl->conns[idx],
+			sizeof(TfwHashConn) * (cl->conn_n - (size_t)idx));
+
+	cl->conns[idx] = new_hcon;
+	++cl->conn_n;
+
+	return 0;
+}
+
+static void
+__fill_srv_lists(TfwHashConnList *cl)
+{
+	size_t i;
 
-	if (!sl)
-		return;
+	for (i = 0; i < cl->conn_n; ++i) {
+		TfwHashConn *hconn = &cl->conns[i];
+		TfwServer *srv = (TfwServer *)hconn->conn->peer;
+		TfwHashConnList *scl = srv->sched_data;
 
-	for (si = 0; si < sl->srv_n; ++si)
-		if (sl->srvs[si].conn)
-			kfree(sl->srvs[si].conn);
-	kfree(sl);
-	sg->sched_data = NULL;
+		scl->conns[scl->conn_n] = *hconn;
+		++scl->conn_n;
+	}
 }
 
 static int
 tfw_sched_hash_add_grp(TfwSrvGroup *sg)
 {
-	int ret = -EINVAL;
-	size_t size, si, ci;
```

Comment:
`si` and `ci` were used to verify that that data in `sg` is self-consistent. That was for protection against possible future changes in `sock_srv.c` by someone who didn't quite realize that `sg` must be completely formed before adding it to a scheduler. This verification is pretty cheap calculation-wise. It is really so unnecessary that it required the removal?